### ISSUE TITLE
temporal filtering before maxwell filtering?

### ISSUE THREAD

**Issue by jasmainak** [mne-tools/mne-study-template#59]
Together with @StevenMGraham we were trying to use the MNE study template for our analysis. 

However, we realized that the study template is doing something different from what the biomag demo repo does. It temporally filters first and then applies maxwell filtering. The biomag demo applies maxwell filtering first.

This seems wrong to me because you'd temporally spread the artifacts before the maxwell filter which would make data overall worse. Instead you want to do  maxwell filtering first and then apply temporal filtering on the cleaned data. Thoughts @agramfort ?

**Comment by agramfort** [comment]
indeed that's what I would recommend

@larsoner agreed?

**Comment by larsoner** [comment]
> This seems wrong to me because you'd temporally spread the artifacts before the maxwell filter which would make data overall worse. Instead you want to do maxwell filtering first and then apply temporal filtering on the cleaned data.

If you are doing tSSS or movement compensation this could matter, because both are basically a time-varying linear operations.

If you are doing plain SSS it should not matter, because vanilla SSS is a spatially linear transformation, and filtering is a temporally linear operation (assuming you apply both operations to all channels under consideration). You should be able to check this on your data.

But yes typically when doing SSS-like operations people do a notch filtering (`filter_chpi`) then SSS(+tSSS+movecomp) then proceed with things like low- or high-pass filtering. The logic behind this is that the notch filtering makes things like tSSS work (probably) better so it's a useful first step, then you proceed with your (potentially) time-varying linear filtering via tSSS+movecomp, then apply your linear temporal filters.

**Comment by virvw** [comment]
My 2 cents while at the Nov hackathon here: filter should be after maxfilter

**Comment by agramfort** [comment]
yes +1 it's more standard

>

**Comment by SophieHerbst** [comment]
We discussed with @agramfort today how to do this. Here is a suggestion: 

- add a new script 00_visualization to plot the raw data and mark bad channels/ segments
- script 01 reads raw and applies maxfilter (sss or tsss) if specified in config, otherwise it saves a _nosss version of raw with bad channels added
- script 02 reads the output of 01 and runs frequency filter with the specified cut-offs 

What do you think @jasmainak @larsoner @virvw ?

I implemented this new order for our version of the pipeline in the team if you want to have a look: [https://github.com/brainthemind/CogBrainDyn_MEG_Pipeline]

**Comment by SophieHerbst** [comment]
PS: there is a known issue [ #6528](https://github.com/mne-tools/mne-python/issues/6528#issuecomment-555536444) with interactive plotting in spyder / jupyter, which is problematic for the above mentioned script 00

**Comment by jasmainak** [comment]
what's the plan with the fork? Do you think we could make a pull request here so that everyone in the community benefits from it?

There is a PR for autobad in Maxfilter in MNE so I am assuming 00_visualization goes away eventually?

The maxfilter in MNE now has autobad IIRC, so maybe we don't need to mark bad channels/segments before?

Also we checked and @larsoner is correct that the order does not matter much at least for the SSS case. However, given that some people might easily swap SSS for tSSS, I think it's safer to keep Maxfilter after temporal filtering.

**Comment by SophieHerbst** [comment]
The fork is too outdated for a mergable PR, but I can make a new PR on the study template. Just wanted to check before doing the work. 

I don't know about autobad, bad for us it is important to be able to look at the data.

**Comment by virvw** [comment]
agrees w/ Sophie - autobad sometimes causes more harm than good


Virginie van Wassenhove
*https://brainthemind.com/ <https://brainthemind.com/>*

**Comment by SophieHerbst** [comment]
I also just realized that we are using a rather specific frequency filter, which comes out as
`one-pass, zero-phase, non-causal`
We should use a two-pass, I think, or put the parameters in the config, not in the script (01 right now).

**Comment by larsoner** [comment]
One-pass FIR should be fine, why do you want two-pass?

**Comment by SophieHerbst** [comment]
to not shift the phase

**Comment by larsoner** [comment]
> zero-phase

This part implies that the phase will not be shifted. You only need to do two-pass with IIR filters to achieve a zero-phase filter. FIR filters can be designed to be zero phase. So no need for two filtering passes to achieve it.

**Comment by SophieHerbst** [comment]
ups, yes. so this shifting back in time preserves the phase-ratio between different frequencies?
Nevertheless, it might make sense to export the filter length, phase, and design parameters to the config file though. 

```
raw.filter(config.l_freq, config.h_freq,
               l_trans_bandwidth=config.l_trans_bandwidth,
               h_trans_bandwidth=config.h_trans_bandwidth,
               filter_length='auto', phase='zero', fir_window='hamming',
               fir_design='firwin'
               )
```

**Comment by jasmainak** [comment]
I believe this should be closed by #63 now?
